### PR TITLE
Implement iterator over database

### PR DESCRIPTION
**Pull request by oschwald**
- Test on 3.12
- Add iterator over Reader object

### CODE REVIEW

File: extension/maxminddb.c [hunk 1]
```diff
@@ -644,13 +857,30 @@ static PyTypeObject Reader_Type = {
...
+static PyTypeObject ReaderIter_Type = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_basicsize = sizeof(ReaderIter_obj),
+    .tp_dealloc = ReaderIter_dealloc,
+    .tp_doc = "Iterator for Reader object",
+    .tp_flags = Py_TPFLAGS_DEFAULT,
+    .tp_iter = PyObject_SelfIter,
+    .tp_iternext = ReaderIter_next,
+    .tp_methods = ReaderIter_methods,
+    .tp_name = "odict_iterator",
```
Comment (horgh) [inline extension/maxminddb.c:880]: Why this name? The docs suggest to me that it should be something specific to this library/type. I'm not sure.

File: extension/maxminddb.c [hunk 2]
```diff
@@ -389,6 +408,200 @@ static void Reader_dealloc(PyObject *self) {
...
+    ReaderIter_obj *ri = PyObject_New(ReaderIter_obj, &ReaderIter_Type);
+    if (ri == NULL) {
+        return NULL;
+    }
+
+    ri->reader = (Reader_obj *)reader;
+    if (ri->reader->closed == Py_True) {
+        PyErr_SetString(PyExc_ValueError,
+                        "Attempt to iterate over a closed MaxMind DB.");
+        return NULL;
```
Comment (horgh) [inline extension/maxminddb.c:421]: Should we be freeing any memory on `ri` in this case? And below when we return null I suppose. Not sure.

File: maxminddb/reader.py [hunk 3]
```diff
@@ -152,6 +166,35 @@ def get_with_prefix_len(
             return self._resolve_data_pointer(pointer), prefix_len
         return None, prefix_len
 
+    def __iter__(self, include_aliased_nodes=False):
```
Comment (horgh) [inline maxminddb/reader.py:173]: It looks like the C version doesn't accept this parameter. Does that matter?

File: extension/maxminddb.c [hunk 4]
```diff
@@ -389,6 +408,200 @@ static void Reader_dealloc(PyObject *self) {
...
+                ri->next = left;
+                break;
+            }
+            case MMDB_RECORD_TYPE_EMPTY:
+                break;
+            case MMDB_RECORD_TYPE_DATA: {
+                MMDB_entry_data_list_s *entry_data_list = NULL;
+                int status =
+                    MMDB_get_entry_data_list(&cur->entry, &entry_data_list);
+                if (MMDB_SUCCESS != status) {
```
Comment (horgh) [inline extension/maxminddb.c:517]: Yoda?

File: maxminddb/reader.py [hunk 5]
```diff
@@ -152,6 +166,35 @@ def get_with_prefix_len(
...
+            # Skip nodes aliased to IPv4
+            return
+
+        node_count = self._metadata.node_count
+        if node > node_count:
+            bits = 128 if self._metadata.ip_version == 6 else 32
+            ip_acc <<= bits - depth
+            if ip_acc <= _IPV4_MAX_NUM and bits == 128:
+                depth -= 96
+            yield ipaddress.ip_network((ip_acc, depth)), self._resolve_data_pointer(
```
Comment (horgh) [inline maxminddb/reader.py:187]: Hmm, the docs don't mention a tuple form of the parameter for ip_network(), do they?

File: extension/maxminddb.c [hunk 6]
```diff
@@ -389,6 +408,200 @@ static void Reader_dealloc(PyObject *self) {
...
+                    Py_BuildValue("(y#i)",
+                                  &(cur->ip_packed[ip_start]),
+                                  ip_length,
+                                  cur->depth - ip_start * 8);
+                if (network_tuple == NULL) {
+                    Py_DECREF(record);
+                    free(cur);
+                    return NULL;
+                }
+                PyObject *args = PyTuple_Pack(1, network_tuple);
```
Comment (horgh) [inline extension/maxminddb.c:557]: Hmm, isn't it already a tuple we could use from the prior call?

**Review by horgh — CHANGES_REQUESTED** [review]
Nice!

File: maxminddb/reader.py (hunk 3 above)
Comment (oschwald) [reply maxminddb/reader.py:173]: Oh, I meant to drop support for this. I'll remove it. I am not sure anyone actually wants the aliased notes and figured I'd leave support for them to a follow-up story.

File: maxminddb/reader.py (hunk 5 above)
Comment (oschwald) [reply maxminddb/reader.py:187]: The parameters are passed through to the constructor and those do mention it. See, e.g., [here](https://docs.python.org/3/library/ipaddress.html#:~:text=A%20two%2Dtuple,g.%20255.255.255.0).).

File: extension/maxminddb.c (hunk 6 above)
Comment (oschwald) [reply extension/maxminddb.c:557]: It is a tuple containing a tuple! :turtle: The args are a tuple and we are passing it a tuple containing the packed address plus the prefix length.

File: extension/maxminddb.c (hunk 2 above)
Comment (oschwald) [reply extension/maxminddb.c:421]: Yeah, good catch. We need to decrement the reference counter.

**Review by horgh — APPROVED** [review]

**Pull request merged**